### ISSUE TITLE
How to visualize the results? 

### ISSUE THREAD

**Issue by DoranLyong** [alzayats/DeepFish#1]
Hello, thanks for sharing your great achievement.  
I also have an interest in counting fish in the water, and I want to use your framework. <br/>

I checked the training code runs in successful (Especially, localization task for counting) <br/>
I want to know how to visualize the results like [this](https://github.com/alzayats/DeepFish/blob/master/docs/count.gif). 

``` python  
# visualize
#model.vis_on_loader(vis_loader, savedir=os.path.join(savedir, "images"))
```
is this code-block for visualizing the result?

**Comment by alzayats** [comment]
Hi, thanks for your feedback.
To visualize the results, predict all the images then use this code to make a movie out of images.

```
import cv2
import os

image_folder = 'images'
video_name = 'video.avi'

images = [img for img in os.listdir(image_folder) if img.endswith(".png")]
frame = cv2.imread(os.path.join(image_folder, images[0]))
height, width, layers = frame.shape

video = cv2.VideoWriter(video_name, 0, 1, (width,height))

for image in images:
    video.write(cv2.imread(os.path.join(image_folder, image)))

cv2.destroyAllWindows()
video.release()
```

**Comment by IssamLaradji** [comment]
Thanks @alzayats , did this solution work for you @DoranLyong ?

**Comment by DoranLyong** [comment]
Thanks for your answers :). <br/> 
In order to visualize the localization task, I'm struggling to make an inference code, but I didn't get any intuition yet. <br/>

For example, I wanted to check the mask data with regard to fish-valid image, so I tried to make both rgb and mask image overlapped like:
```python 
val_set = datasets.get_dataset(dataset_name=exp_dict["dataset"], split="val",
                                   transform=exp_dict.get("transform"),
                                   datadir=datadir)

plt.imshow(val_set[150]['images'].permute(1,2,0))
plt.imshow(val_set[150]['points'], alpha=0.80, cmap='gray')
```
But, it seems hard to check like this: 
![image](https://user-images.githubusercontent.com/13333380/100519577-32378e80-31dc-11eb-86ae-05264d1e4402.png)

and I'm meeting some problems to make an inference code following your training code. <br/>
Could you please suggest a guideline prediction code for localization?

**Comment by alzayats** [comment]
Hi @DoranLyong ,

To check the mask data with regard to fish-valid image please use this code from [blobs_functions.py](https://github.com/alzayats/DeepFish/blob/master/src/modules/blobs_functions.py)

```
'''all_visualize_dots'''
def all_visualize_dots():
    from os.path import splitext
    imgs_path = 'D:/Datasets/JCU_FISH/jcu_fish_step10_1617/images/valid/'
    points_path = 'D:/Datasets/JCU_FISH/jcu_fish_step10_1617/masks/valid/'
    root_path = 'D:/Datasets/JCU_FISH/jcu_fish_train_yes_step10_valid_png/visualize_dots/'
    for file in tqdm(listdir(imgs_path)):
        img_path = imgs_path + file
        point_path = points_path + splitext(file)[0] + '.png'
        save_path = root_path + 'dots_' +file
        visualize_dots(img_path, point_path, save_path, size=10)

```

BTW, your code did not visualize the annotation because it is a point annotation (i.e. just one pixel per fish, so you can not see one pixel easily). Our code covert the (one pixel) to a circle.

`Could you please suggest a guideline prediction code for localization?`
Please use `vis_on_batch` function that overlays the LCFCN prediction on the input image from[ loc_wrapper.py](https://github.com/alzayats/DeepFish/blob/ea77a1bdfc2a72af0d66741d59b99389f0627184/src/wrappers/loc_wrapper.py)

I hope this answers your questions.

Cheers,

**Comment by DoranLyong** [comment]
Thank you for your kindness! Yes, I find out that part yesterday and got a ```predicted mask```, but the results were no good. <br/> Maybe because I didn't train with 300 epochs, not 1000 epochs as you set? So, I'm just training with the default setup as in [exp_configs.py](https://github.com/alzayats/DeepFish/blob/master/exp_configs.py) for ```"loc" ```.  However, the ```loss_loc``` seems not that much decreasing yet.  <br/>

Could I get some pre-trained weight file for checking your suggestion?

**Comment by BJTUJia** [comment]
@IssamLaradji @alzayats Hi there, I have the same problem with the segmentation task. I used your codes and configs to train the segmentation task, after which I want to visualize the results. I simply use the vis_on_batch function in seg_wrapper.py, but the output results are weird, the fishes are not marked at all, like this. Can you kindly help me figure out what might be causing this issue. Thanks!
![image](https://user-images.githubusercontent.com/24791613/101305653-1e47f880-387e-11eb-96dd-aa4b6a861181.png)

**Comment by BJTUJia** [comment]
![image](https://user-images.githubusercontent.com/24791613/101306189-2c4a4900-387f-11eb-9035-c301619b4ce3.png)

**Comment by IssamLaradji** [comment]
Hi @BJTUJia , I have fixed the segmentation issues, the reason why you weren't getting good results is because the default was not an ImageNet pretrained network. Since that has been fixed, you can test the following,

0) Do a `git pull`

1) Please install the following `pip install git+https://github.com/ElementAI/LCFCN`

2) Then run 

```
python scripts/train_single_image.py -e seg -d ${PATH_TO_DATASET}
```

you should get something like this,

![image](https://user-images.githubusercontent.com/3382128/102641236-bd47dc80-4129-11eb-8312-ac0bc33371e9.png)


Left hand side is the ground truth,  right is the predicted segmentation of the network. Let me know if you face issues.

**Comment by BJTUJia** [comment]
Thanks!  I will try it out!

**Comment by IssamLaradji** [comment]
you welcome! :))
